Add type and instance filters for hypotheses

lovettchris opened the pull request:
Part II of issue https://github.com/leanprover/vscode-lean4/issues/76  is adding filter buttons for hiding hypotheses that have either isType or isInstance set to true.  

https://user-images.githubusercontent.com/18707114/169914051-a802e4b9-5fa4-4fc6-9d7c-34ed6af91517.mp4

gebner commented on lean4-infoview/src/infoview/rpcInterface.ts:
```diff
@@ -65,6 +65,8 @@ export async function InteractiveDiagnostics_infoToInteractive(rs: RpcSessions,
 }
 
 export interface InteractiveHypothesis {
+    isInstance: boolean,
+    isType: boolean,
```
```suggestion
    isInstance?: boolean,
    isType?: boolean,
```
Since this property is not present in older Lean versions.

lovettchris replied:
thanks

gebner commented on lean4-infoview/src/infoview/goals.tsx:
```diff
@@ -29,38 +29,60 @@ export function goalsToString(goals: InteractiveGoals): string {
+}
+
+function getFilteredHypotheses(hyps: InteractiveHypothesis[], filter: GoalFilterState): InteractiveHypothesis[] {
+    const result : InteractiveHypothesis[] = [];
+    for (const h of hyps) {
+        if (h.isInstance) {
+            if (filter.isInstance)
+                result.push(h);
+        }
+        else if (h.isType) {
+            if (filter.isType)
+                result.push(h);
+        }
+        else{
+            // TODO: are there other types?
+            result.push(h);
+        }
+    }
+    return result;
```
```suggestion
    return hyps.filter(h =>
        (!h.isInstance || filter.isInstance) &&
        (!h.isType || filter.isType));
```

gebner commented on lean4-infoview/src/infoview/info.tsx:
```diff
@@ -123,8 +124,37 @@ export function InfoDisplay(props0: InfoDisplayProps) {
-    const sortButton = <a className={filterClasses} onClick={e => { setReverseOrder(!reverseOrder); }} title="reverse list"/>
+    const sortClasses = 'link pointer mh2 dim codicon fr ' + (goalFilters.reverse ? 'codicon-arrow-up' : 'codicon-arrow-down');
+    const sortButton = <a className={sortClasses} title="reverse list" onClick={e => {
+        setGoalFilters(s => {
+            return { reverse: !s.reverse, isType: s.isType, isInstance: s.isInstance }
+        } ); }
```
```suggestion
        setGoalFilters(s => ({ ...s, reverse: !s.reverse })); }
```

lovettchris replied:
Thanks, I fixed all 3 places this was needed.

gebner commented on lean4-infoview/src/infoview/info.tsx:
```diff
@@ -123,8 +124,37 @@ export function InfoDisplay(props0: InfoDisplayProps) {
+            setGoalFilters(s => {
+                return { reverse: s.reverse, isType: !s.isType, isInstance: s.isInstance }
+            } ); }}>
+                <span className={'filterMenuIcon codicon ' + (goalFilters.isType ? 'codicon-eye' : 'codicon-eye-closed')}>&nbsp;</span>
```
The "eye" codicon looks very cryptic to me.  Maybe it would be more intuitive if this was an actual checkbox, and the text was "show instances" (or "hide instances").

lovettchris replied:
Ok, I found another filter menu in the Problems UI that does use checkboxes, so consistency wins:
![image](https://user-images.githubusercontent.com/18707114/169890006-78cd2e8b-f517-4488-964b-5f2524bd9757.png)

Even though this is exactly what the eye icon was invented for :-) It literally means "show it so I can see it" or "hide it so I cannot see it", but I guess not enough apps have used it over the years which results in your reaction...

Problem is now I have to find a "blank filler" that is exactly the same size as the checkmark icon but codeicons has no such blank icon defined...

lovettchris replied:
I can't use a fixed pixel sized filler because it changes with vs code font size (Ctrl + and Ctrl -)

gebner replied:
I understand the eye icon in general, it's just that the codicons one doesn't look at all like an eye at all.  The checkmark is certainly clearer (at least to me).

> Problem is now I have to find a "blank filler" that is exactly the same size as the checkmark icon but codeicons has no such blank icon defined...

The two eye icons also had different sizes, it's not the end of the world if they don't match exactly.

lovettchris replied:
Turns out fixed pixel icon does work, the Ctrl+ key must be scaling the DPI.  I updated the demo video to show:

![image](https://user-images.githubusercontent.com/18707114/169914292-8db9a100-e446-44a0-b1d7-e5ab58228c59.png)

lovettchris replied:
Note: I did find one eye icon usage in vscode for lazily expanded variables in the debugger:
![image](https://user-images.githubusercontent.com/18707114/169933458-ba99b3d2-ce47-4650-9218-c700ccec7151.png)

gebner commented on lean4-infoview/src/infoview/goalCompat.ts:
```diff
@@ -31,6 +31,8 @@ function transformGoalToInteractive(g: string): InteractiveGoal {
         } else if (p.match(/^([^:\n< ][^:\n⊢{[(⦃]*) :/mg)) {
             const ss = p.split(':')
             const hyp: InteractiveHypothesis = {
+                isType: false,
+                isInstance: false,
```
```suggestion
                isType: true,
                isInstance: true,
```

We shouldn't hide anything by default (especially when there's no button to change the filtering).

gebner replied:
Oh then it's ok.  Got the flags backwards.

gebner replied:
I actually confused the filter flags with the hypothesis flags. :bulb:

lovettchris replied:
Ah but I did find another place in traceExplorer.tsx that needed a related fix.

lovettchris commented:
While I'm at it, I added hidden assumption filter based on presence of the dagger in the hyp name since you asked for that in issue# 76 as follows:
![image](https://user-images.githubusercontent.com/18707114/169920698-06a4a451-7fe9-4d5c-ba81-dc25a3955dbc.png)

which filters down to:
![image](https://user-images.githubusercontent.com/18707114/169920711-63887499-965c-44f8-9e11-7514f3958188.png)

gebner commented:
Works great!  Two minor nitpicks:
 1. The contrast between the links and the tooltip background is pretty bad with light themes.
 2. If only hidden assumptions is checked, then the filter icon is not "filled".

![filters_lightmode](https://user-images.githubusercontent.com/313929/170107188-2c88648a-4d0f-435c-b483-892219d9c7cc.png)

lovettchris commented:
> Works great! Two minor nitpicks:
> 1. The contrast between the links and the tooltip background is pretty bad with light themes.
> 2. If only hidden assumptions is checked, then the filter icon is not "filled".

Good catch thanks, we also have a contrast problem in regular tooltips with the white on light blue shown here:
![image](https://user-images.githubusercontent.com/18707114/170119797-ad5d0419-11bc-43a4-9f11-d68e5dd8193c.png)

I pushed a fix so the light theme now looks like this:
![image](https://user-images.githubusercontent.com/18707114/170123686-d7765250-bb78-4ed8-b8a9-edf2e73aa5f6.png)

And the dark theme like this:
![image](https://user-images.githubusercontent.com/18707114/170123877-add456f4-5975-44b9-be37-457af3d9c7e0.png)

I also fixed the filter icon state bug you found.

gebner commented:
I like the white background in the tooltips, but the text should be black (for contrast; the same color as the text outside the tooltip).

@Vtec234 When you added the tooltips originally, you made them black even in light themes (but not white in dark themes).  (There's even a comment: "tooltips are _always_ dark.")  Did you have a reason for that?

lovettchris commented:
> I like the white background in the tooltips, but the text should be black (for contrast; the same color as the text outside the tooltip).

As you can see in index.css I'm using the following official VS code themed colors for editor widgets, so I'm inclined to stick with that:
```
    color: var(--vscode-editorWidget-foreground);
    background: var(--vscode-editorWidget-background);
```

Vtec234 commented:
>  When you added the tooltips originally, you made them black even in light themes (but not white in dark themes). (There's even a comment: "tooltips are always dark.") Did you have a reason for that?

There.. is? :) You're right, and I don't remember why. Chris's changes look much better, let's go with those. I think for the dark-on-dark tooltips, though, we should change the shadow colour to a light one, if possible. In light themes the dark shadow makes it easy to see the tooltip boundary whereas in dark themes a dark shadow doesn't.

> As you can see in index.css I'm using the following official VS code themed colors for editor widgets, so I'm inclined to stick with that:

Note though that in your screenshot, in the light theme, text inside the tooltip is lighter than other infoview text (unless that's an illusion?).

Vtec234 commented on lean4-infoview/src/infoview/goals.tsx:
```diff
@@ -29,38 +29,53 @@ export function goalsToString(goals: InteractiveGoals): string {
 }
 
-export function Goal({pos, goal, reverse}: {pos: DocumentPosition, goal: InteractiveGoal, reverse: boolean}) {
+export interface GoalFilterState {
```
Let's put a comment on the interface as to which way the `isFoo` flags go. Does `true` mean show or hide?

lovettchris replied:
Done

Vtec234 reviewed (approved):
Looks great!

Vtec234 commented:
As another visual suggestion, I would remove the `HighlightOnHoverSpan`. It doesn't align with the icon ![screenshot](https://user-images.githubusercontent.com/13901751/170331260-1fe4f925-a8bf-4a98-98cb-f8e0c4d07abb.png), and I think the `dim` effect is enough of an indication for 'here is something you can interact with'.

lovettchris commented:
> we should change the shadow colour to a light one

Turns out that VS code does define `--vscode-widget-shadow` and they've decided to make it black in the dark theme for some reason.  I switched to that for now.

> I like the white background in the tooltips, but the text should be black (for contrast; the same color as the text outside the tooltip).

I switched from `color: var(--vscode-editorWidget-foreground);` to `@extend .mtk1; which is the color used by the text editor, so the text is now black on light theme :
![image](https://user-images.githubusercontent.com/18707114/170335468-35b9d2e0-f22a-42fd-b790-7a8e9a8ecf4a.png)
and white on dark theme:
![image](https://user-images.githubusercontent.com/18707114/170335692-60b4d34a-9381-4dea-a19a-628afa09514b.png)

lovettchris commented:
> As another visual suggestion, I would remove the `HighlightOnHoverSpan`. It doesn't align with the icon ![screenshot](https://user-images.githubusercontent.com/13901751/170331260-1fe4f925-a8bf-4a98-98cb-f8e0c4d07abb.png), and I think the `dim` effect is enough of an indication for 'here is something you can interact with'.

Done.

Vtec234 commented on lean4-infoview/src/infoview/index.css:
```diff
@@ -118,3 +121,17 @@ select {
     left: -4px;
 }
+
+.popup-menu-text { text-align: center; padding: 5px; vertical-align:top; display:inline-block; }
```
I missed this before, but so far we have mostly been using [Tachyons](http://tachyons.io/docs/table-of-styles/) rather than bespoke CSS classes for styling of this kind (with the exception of this exact file, of course). So instead of `popup-menu-text` one would write `tc pa2 v-top dib` as the `className` (`pa1` or `pa2`, Tachyons intentionally doesn't provide specific pixel offsets for padding so as to keep a consistent, fixed scale). All the other classes either use `var` or *need* specific offsets to match a font, so it only applies to `popup-menu-text`. I will leave to you whether you want to change this, it's not the end of the world if we use classes.

lovettchris replied:
Interesting, I have lots of questions, so I moved this to a new issue  https://github.com/leanprover/vscode-lean4/issues/183

lovettchris commented:
I went through all the comments again, I think this is done.  Did I miss anything?

Vtec234 commented:
Merged with history since this is a large change, but please do follow the lean4-like [commit convention](https://github.com/leanprover/lean4/blob/master/doc/dev/commit_convention.md).

lovettchris commented:
> Merged with history since this is a large change, but please do follow the lean4-like [commit convention](https://github.com/leanprover/lean4/blob/master/doc/dev/commit_convention.md).

woops, yeah, I assumed it would be squash merged...
